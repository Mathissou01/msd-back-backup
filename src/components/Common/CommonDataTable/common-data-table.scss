@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.shadows" as shadows;
@use "src/styles/02-tools/tools.focus" as focus;

.c-CommonDataTable {
  @include shadows.shadow(1);
  @include focus.focus-ring(true);

  border-radius: 8px 8px 0 0 !important;

  // stylelint-disable selector-class-pattern
  .rdt_TableHead,
  .rdt_TableRow {
    color: colors.$expert-blue;
    font-weight: 500;
    font-size: 16px;
    font-family: DINPro-Medium, sans-serif;
    line-height: 24px;
  }

  .rdt_TableHeadRow {
    padding: 24px 0 15px;
    background-color: colors.$blue-light;
  }

  .rdt_TableCol,
  .rdt_TableCell {
    padding-right: 12px;
    padding-left: 12px;
  }

  .rdt_TableCol_Sortable {
    gap: 10px;

    span {
      content: url("data:image/svg+xml,%3Csvg width='11' height='7' viewBox='0 0 11 7' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M5.242 3.828 8.777.293a1 1 0 0 1 1.415 1.414L5.948 5.95a1 1 0 0 1-1.415-.001L.292 1.707A1 1 0 0 1 1.706.293l3.536 3.535Z' fill='%23030F40'/%3E%3C/svg%3E");
    }
  }

  .rdt_TableRow {
    min-height: 60px;
    padding: 12px 0;
    line-height: 24px;

    &:not(:last-of-type) {
      border-bottom-color: colors.$blue-4;
    }
  }

  // stylelint-enable selector-class-pattern

  &_disabled {
    opacity: 0.5;
    pointer-events: none;
  }
}
