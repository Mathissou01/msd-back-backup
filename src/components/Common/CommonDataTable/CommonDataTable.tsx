import classNames from "classnames";
import DataTable, {
  PaginationComponentProps,
  SortOrder,
  TableColumn,
} from "react-data-table-component";
import React, { useEffect, useRef, useState } from "react";
import { removeNulls } from "../../../lib/utilities";
import DataTableFilters, {
  IDataTableFilter,
} from "./DataTableFilters/DataTableFilters";
import CommonPagination from "../CommonPagination/CommonPagination";
import CommonSpinner from "../CommonSpinner/CommonSpinner";
import DataTableActions, {
  IDataTableAction,
} from "./DataTableActions/DataTableActions";
import "./common-data-table.scss";

export interface ICommonDataTableValidation {
  isValid: boolean;
  errorMessage?: string;
}

export interface IDefaultTableRow {
  id: string;
  editState: boolean;
}

interface ILazyLoadingOptions {
  isRemote: boolean;
  totalRows: number;
}

interface IPaginationOptions {
  hasPagination: boolean;
  hasRowsPerPageOptions?: boolean;
  defaultRowsPerPage: number;
  rowsPerPageOptions?: Array<number>;
  defaultPage?: number;
}

export interface ISortingParams {
  column: string;
  direction?: "asc" | "desc";
}

export interface ICurrentPagination<T> {
  page: number;
  rowsPerPage: number;
  filter?: IDataTableFilter<T>;
  sort?: ISortingParams;
}

interface ICommonDataTableProps<T> {
  columns: Array<TableColumn<T>>;
  actionColumn?: (row: T, rowIndex: number) => Array<IDataTableAction>;
  data: Array<T>;
  lazyLoadingOptions?: ILazyLoadingOptions;
  // State
  isLoading?: boolean;
  // Filters
  filters?: Array<IDataTableFilter<T>>;
  // Sorting
  defaultSortFieldId?: string | number;
  // Pagination
  paginationOptions?: IPaginationOptions;
  onLazyLoad?: (params: ICurrentPagination<T>) => void;
}

export default function CommonDataTable<T extends IDefaultTableRow>({
  columns,
  actionColumn,
  data,
  lazyLoadingOptions,
  isLoading,
  filters,
  defaultSortFieldId,
  paginationOptions,
  onLazyLoad,
}: ICommonDataTableProps<T>) {
  /* Static Data */
  const noRecordsLabel = "Il n'y a pas de données à afficher";
  const noRecords = (
    <div className="c-CommonDataTable__NoRecords">
      <span>{noRecordsLabel}</span>
    </div>
  );

  /* Methods */
  function handleFilter(filter: IDataTableFilter<T>) {
    if (lazyLoadingOptions?.isRemote) {
      setCurrentPagination({ ...currentPagination, filter });
      setResetDefaultPage(true);
    } else {
      setTableData(data.filter((row) => filter.selector?.(row) ?? true));
    }
  }

  function handleSort(
    selectedColumn: TableColumn<T>,
    sortDirection: SortOrder,
  ) {
    if (lazyLoadingOptions?.isRemote && selectedColumn.id) {
      setCurrentPagination({
        ...currentPagination,
        sort: {
          column: selectedColumn.id?.toString(),
          direction: sortDirection,
        },
      });
      setResetDefaultPage(true);
    }
  }

  /* Local Data */
  const [tableData, setTableData] = useState<Array<T>>(data);
  const [filterTabs, setFilterTabs] = useState<Array<IDataTableFilter<T>>>([]);
  const [resetDefaultPage, setResetDefaultPage] = useState(false);
  const [currentPagination, setCurrentPagination] = useState<
    ICurrentPagination<T>
  >({
    page: paginationOptions?.defaultPage ?? 1,
    rowsPerPage: paginationOptions?.defaultRowsPerPage ?? 10,
  });
  const prevPagination = useRef<ICurrentPagination<T>>(currentPagination);
  const tableColumns: Array<TableColumn<T>> = [
    ...columns,
    ...[
      actionColumn && actionColumn.length > 0
        ? {
            id: "actions",
            name: "",
            cell: (row: T, rowIndex: number) => {
              const tableActions = actionColumn(row, rowIndex);
              return <DataTableActions actions={tableActions} />;
            },
            width: `${36 * 3 + 32}px`,
            grow: 0,
            center: true,
          }
        : {},
    ],
  ].filter(removeNulls);

  // Parent data change
  useEffect(() => {
    setTableData(data);
    setResetDefaultPage(false);
  }, [data]);

  // Filters change
  useEffect(() => {
    if (tableData && tableData.length > 0 && filters && filters.length > 0) {
      setFilterTabs(
        lazyLoadingOptions?.isRemote
          ? filters
          : filters?.map((filter) => {
              return {
                ...filter,
                count:
                  filter.count ?? filter.selector
                    ? data.filter((row) => filter.selector?.(row)).length
                    : null,
              };
            }),
      );
    }
  }, [data, filters, lazyLoadingOptions?.isRemote, tableData]);

  // Pagination change
  useEffect(() => {
    if (
      lazyLoadingOptions?.isRemote &&
      currentPagination !== prevPagination.current
    ) {
      prevPagination.current = currentPagination;
      onLazyLoad?.({ ...currentPagination });
    }
  }, [lazyLoadingOptions?.isRemote, currentPagination, onLazyLoad]);

  const CustomPagination = ({
    rowsPerPage,
    rowCount,
    onChangePage,
    onChangeRowsPerPage,
    currentPage,
  }: PaginationComponentProps) => (
    <div className="c-CommonDataTable__Pagination">
      <CommonPagination
        currentPage={currentPage}
        rowCount={rowCount}
        onChangePage={(page, totalRows) => onChangePage(page, totalRows)}
        onChangeRowsPerPage={(currentRowsPerPage, currentPage) =>
          onChangeRowsPerPage(currentRowsPerPage, currentPage)
        }
        rowsPerPage={rowsPerPage}
        {...(!!paginationOptions && {
          noRowsPerPage: !(paginationOptions.hasRowsPerPageOptions ?? true),
          rowsPerPageOptions: paginationOptions.rowsPerPageOptions,
        })}
      />
    </div>
  );

  return (
    <>
      {filterTabs && filterTabs.length > 0 && (
        <DataTableFilters<T>
          tabs={filterTabs}
          onFilter={(filter) => handleFilter(filter)}
        />
      )}
      <div
        className={classNames("c-CommonDataTable", {
          "c-CommonDataTable_loading": isLoading,
        })}
      >
        <DataTable<T>
          columns={tableColumns}
          data={tableData}
          progressPending={isLoading}
          progressComponent={<CommonSpinner />}
          noDataComponent={noRecords}
          defaultSortFieldId={defaultSortFieldId}
          defaultSortAsc={true}
          onSort={(selectedColumn, sortDirection) =>
            handleSort(selectedColumn, sortDirection)
          }
          {...(paginationOptions &&
            paginationOptions.hasPagination && {
              pagination: true,
              paginationPerPage: paginationOptions.defaultRowsPerPage,
              paginationRowsPerPageOptions:
                paginationOptions.rowsPerPageOptions,
              paginationComponent: CustomPagination,
              paginationResetDefaultPage: resetDefaultPage,
              onChangePage: (currentPage) => {
                if (currentPage !== currentPagination.page) {
                  setCurrentPagination({
                    ...currentPagination,
                    page: currentPage,
                  });
                }
              },
              onChangeRowsPerPage: (currentRowsPerPage) => {
                if (currentRowsPerPage !== currentPagination.rowsPerPage) {
                  setCurrentPagination({
                    ...currentPagination,
                    rowsPerPage: currentRowsPerPage,
                  });
                }
              },
            })}
          {...(lazyLoadingOptions?.isRemote && {
            sortServer: true,
            paginationServer: true,
            paginationTotalRows: lazyLoadingOptions.totalRows,
          })}
        />
      </div>
    </>
  );
}
