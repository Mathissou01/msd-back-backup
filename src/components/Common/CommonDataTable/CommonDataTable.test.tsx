import { render, screen, within } from "@testing-library/react";
import React from "react";
import CommonDataTable from "./CommonDataTable";
import "jest-styled-components";

interface ITestRow {
  id: string;
  editState: boolean;
  name: string;
}

const mocks = {
  columns: [
    {
      id: "id",
      selector: (row: ITestRow) => row.id,
      omit: true,
    },
    {
      id: "name",
      name: "name",
      selector: (row: ITestRow) => row.name,
    },
  ],
  data: [
    { id: "1", editState: false, name: "first name" },
    { id: "2", editState: false, name: "second name" },
    { id: "5", editState: false, name: "fifth name" },
    { id: "10", editState: false, name: "tenth name" },
  ],
};

describe("CommonDataTable", () => {
  it("renders", () => {
    const { container } = render(
      <CommonDataTable columns={mocks.columns} data={mocks.data} />,
    );

    const table = screen.getByRole("table");
    expect(table).toBeInTheDocument();
    const rowGroups = within(table).getAllByRole("rowgroup");
    const rows = within(table).getAllByRole("row");
    expect(rowGroups.length).toBe(2);
    expect(rows.length).toBe(5);
    expect(container).toMatchSnapshot();
  });
});
