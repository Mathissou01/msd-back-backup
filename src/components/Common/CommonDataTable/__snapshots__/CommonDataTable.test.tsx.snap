// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonDataTable renders 1`] = `
<div>
  <div
    class="c-CommonDataTable"
  >
    <div
      class="sc-dmXXqC dLPHLX"
    >
      <div
        class="sc-fLsdZI fOcRrp"
      >
        <div
          class="sc-beySbM dehtZj rdt_Table"
          role="table"
        >
          <div
            class="sc-guDLRT jnKa-Dc rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="sc-dmyDGi eYOasi rdt_TableHeadRow"
              role="row"
            >
              <div
                class="sc-hLQRIN sc-eDLJxc sc-irLwvL hyxmwx lmTtlo ftBQWI rdt_TableCol"
                data-column-id="name"
              >
                <div
                  class="sc-csKIKE dQRpGB rdt_TableCol_Sortable"
                  data-column-id="name"
                  data-sort-id="name"
                  disabled=""
                  role="columnheader"
                  tabindex="0"
                >
                  <div
                    class="sc-eTNep JRhAO"
                    data-column-id="name"
                  >
                    name
                  </div>
                </div>
              </div>
              <div
                class="sc-hLQRIN sc-eDLJxc sc-irLwvL hyxmwx lmTtlo ftBQWI rdt_TableCol"
                data-column-id="3"
              />
            </div>
          </div>
          <div
            class="sc-hIPCAL dtiYrZ rdt_TableBody"
            role="rowgroup"
          >
            <div
              class="sc-jsEegt jtfTQm rdt_TableRow"
              id="row-1"
              role="row"
            >
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-1"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  first name
                </div>
              </div>
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="3"
                data-tag="allowRowEvents"
                id="cell-3-1"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                />
              </div>
            </div>
            <div
              class="sc-jsEegt jtfTQm rdt_TableRow"
              id="row-2"
              role="row"
            >
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-2"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  second name
                </div>
              </div>
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="3"
                data-tag="allowRowEvents"
                id="cell-3-2"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                />
              </div>
            </div>
            <div
              class="sc-jsEegt jtfTQm rdt_TableRow"
              id="row-5"
              role="row"
            >
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-5"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  fifth name
                </div>
              </div>
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="3"
                data-tag="allowRowEvents"
                id="cell-3-5"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                />
              </div>
            </div>
            <div
              class="sc-jsEegt jtfTQm rdt_TableRow"
              id="row-10"
              role="row"
            >
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-10"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  tenth name
                </div>
              </div>
              <div
                class="sc-hLQRIN sc-eDLJxc sc-jTQDnj kDZMuE lmTtlo chdMBN rdt_TableCell"
                data-column-id="3"
                data-tag="allowRowEvents"
                id="cell-3-10"
                role="cell"
              >
                <div
                  data-tag="allowRowEvents"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
